### ISSUE TITLE
Stack injected packets are also received on rings in rx mode

### ISSUE THREAD

**Issue by Eichhoernchen** [ntop/PF_RING#27]
When using PF_RING and listening in ZC mode (so no packet reach the kernel) and packets that are injected via stack:ethX into the kernel the pfring listening on zc:ethX will receive those packets.

Depending on the application this may result in infinite loops.

E.g.: 
Your tool listens on zc:eth1 and in case there is something that absolutely hast to get into the kernel uses stack:eth1 to push it there then it will receive the same packet it just pushed on zc:eth1 as well, repeat :)

One could easily solve this by (ab)using the sb_buff cb field when creating an skb that is to be locally injected and checking for said field in the packet function hook (packet_rcv). 

I patched this locally, if you feel that this is not the way injected packets should be handled I can submit a patch.

**Comment by cardigliano** [comment]
This is unexpected, which driver are you using?
Please also submit the patch for review.

Thank you
Alfredo

**Comment by Eichhoernchen** [comment]
I'm on the head of the dev branch 

PF_RING Version          : 6.1.1 (dev:315849654f2f98344b7d37898a256c3c3317f9c0)


You can simply try:

zcount -i zc:ethX -c 11

and

echo "ffffffffffff0026444fb191080600010800060400010026444fb191c0a801fe000000000000a9fecbe1" | pfsend -i stack:ethX -n1000

You will see that zcount will see 1000 packets which I think it should not see.


Here is a patch:

```
diff --git a/kernel/pf_ring.c b/kernel/pf_ring.c
index 488623d..7d4df13 100644
--- a/kernel/pf_ring.c
+++ b/kernel/pf_ring.c
@@ -4786,15 +4786,20 @@ static int packet_rcv(struct sk_buff *skb, struct net_device *dev,
   int rc;
   u_int8_t skb_reference_in_use = 0;

-  if(skb->pkt_type != PACKET_LOOPBACK) {
-    rc = skb_ring_handler(skb,
+  if(skb->cb[0] != 1) {
+    if(skb->pkt_type != PACKET_LOOPBACK) {
+      rc = skb_ring_handler(skb,
                          (skb->pkt_type == PACKET_OUTGOING) ? 0 : 1,
                          1 /* real_skb */, &skb_reference_in_use,
                          -1 /* unknown: any channel */,
                           UNKNOWN_NUM_RX_CHANNELS);

-  } else
-    rc = 0;
+    } else {
+      rc = 0;
+    }
+  }else {
+     rc = 0;
+  }

   if(!skb_reference_in_use)
     kfree_skb(skb);
@@ -6490,6 +6495,7 @@ static int pf_ring_inject_packet_to_stack(struct net_device *netdev, struct msgh
   if(err)
     return err;
   skb->protocol = eth_type_trans(skb, netdev);
+  skb->cb[0] = 1;
   err = netif_rx_ni(skb);
   if (unlikely(enable_debug && err == NET_RX_SUCCESS))
     printk("[PF_RING] Packet injected into the linux kernel!\n");
```

**Comment by cardigliano** [comment]
@Eichhoernchen what is the driver model you are using? (ixgbe, igb, e1000e)

**Comment by Eichhoernchen** [comment]
e1000e

**Comment by Eichhoernchen** [comment]
So... the patch is not really working :) as there are outgoing packets that have the cb set e.g. by tcp...

However after more digging I saw that in `skb_ring_handler` there is actually a check for `is_valid_skb_direction`... However I have the feeling that the different sockets aren't configured correctly.

I use the following code: 
```

zc_from_card = pfring_zc_open_device(cluster, zc_eth, rx_only, 0);
zc_to_card = pfring_zc_open_device(cluster, zc_eth, tx_only, 0);

dev_to_kernel = pfring_zc_open_device(cluster, stack_eth, tx_only, 0);
dev_from_kernel = pfring_zc_open_device(cluster, stack_eth, rx_only, 0);

```
If I write packets to the kernel on dev_to_kernel, I will get them on zc_from_card...

What I basically want is zc_from_card -> dev_to_kernel -> dev_from_kernel -> zc_to_card
( In very rare events)

**Comment by coelho** [comment]
I can confirm this bug. zbounce between zc:eth0 and stack:eth0 (for example) works perfectly fine if it runs on ixgbe, but gets a ton of duplicate packets on e1000e.
It's a bit frustrating during testing.

```
# ping 10.0.5.12
PING 10.0.5.12 (10.0.5.12) 56(84) bytes of data.
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.319 ms
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.418 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.472 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.525 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.578 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.632 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.685 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.739 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.793 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.846 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.899 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=0.953 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=1.00 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=1.06 ms (DUP!)
64 bytes from 10.0.5.12: icmp_seq=63 ttl=64 time=1.11 ms (DUP!)
```

**Comment by cardigliano** [comment]
I will look at this, thank you for reporting.

Alfredo

**Comment by cardigliano** [comment]
Please check latest driver from git dev, it should also fix this. Thank you.

**Comment by cardigliano** [comment]
Fixed with new driver e1000e-3.2.4.2-zc and commits below:
https://github.com/ntop/PF_RING/commit/6f7e3123969b50f05cafdd64807a2e9522d167d6
https://github.com/ntop/PF_RING/commit/fbe98a4465f7a51a7543396c02adf26c8fb45091
